### PR TITLE
Feature/growth checker

### PR DESCRIPTION
**Pull request by rycastaneda**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,9 +3,13 @@
...
   "dependencies": {
+    "anytv-node-mysql": "^0.0.13",
+    "async": "^1.2.1",
+    "cuddle": "0.0.48",
     "flatiron": "0.4.3",
-    "mysql": "2.6.1",
-    "anytv-node-mysql": "^0.0.13"
+    "moment": "^2.10.3",
+    "moment-range": "^2.0.2",
+    "mysql": "2.6.1"
```
Comment (eshleebien) [inline package.json]: you remove this

File: commands/growth_checker.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        },
+
+        process_rows = function (err, results) {
+            if (err) {
+                app.log.error('Something went wrong.');
+                return next();
+            }
+            // console.log("results.length",results.length);
+            async.map(results, get_youtube, function (err, result) {
+                app.log.info("result",result);
```
Comment (eshleebien) [inline commands/growth_checker.js]: space after comma

File: commands/growth_checker.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        get_channels = function () {
+            var query = 'SELECT channel_id, network FROM mcn_channels WHERE (linked_date != "" OR linked_date is not null) AND (network != "")';
+
+            if (app.argv.limit && app.argv.limit > 0) {
+                limit = app.argv.limit;
+            }
+
+            query += 'LIMIT ' + limit;
+
+            if (app.argv.offset && app.argv.offset > 0) {
```
Comment (pprmint) [inline commands/growth_checker.js]: `+app.argv.offset`

File: commands/growth_checker.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+            return mysql.open(app.config.get('database').heartbeat)
+                .query(query, process_rows);
+        },
+
+        process_rows = function (err, results) {
+            if (err) {
+                app.log.error('Something went wrong.');
+                return next();
+            }
```
Comment (adin234) [inline commands/growth_checker.js]: we highly suggest putting new lines after `}`

File: commands/growth_checker.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            async.map(results, get_youtube, function (err, result) {
+                app.log.info("result", result);
+                app.log.info('successful: ', success);
+                app.log.info('failed: ', failed);
```
Comment (adin234) [inline commands/growth_checker.js]: do we really want to log in an info level the failures?

File: commands/growth_checker.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+                    part: 'snippet',
+                    id : channel.channel_id,
+                    fields: 'items/snippet',
+                    key: app.config.get('google_api')[0].key
+                }).then(function (err, data) {
+                    if (err) {
+                        app.log.error(channel.channel_id + ' cannot find youtube data.');
+                        failed++;
+                        return cb(null, {channel_id : channel.channel_id, error: 'cannot find youtube data'});
+                    }
```
Comment (adin234) [inline commands/growth_checker.js]: we highly suggest putting new lines after `}`

File: commands/growth_checker.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                .send({
+                    part: 'snippet',
+                    id : channel.channel_id,
+                    fields: 'items/snippet',
+                    key: app.config.get('google_api')[0].key
+                }).then(function (err, data) {
+                    if (err) {
+                        app.log.error(channel.channel_id + ' cannot find youtube data.');
+                        failed++;
+                        return cb(null, {channel_id : channel.channel_id, error: 'cannot find youtube data'});
```
Comment (adin234) [inline commands/growth_checker.js]: don't we want to call `cb` here sending an error?

File: commands/growth_checker.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                    }
+                    return get_dates(data, channel, cb);
+                })
```
Comment (adin234) [inline commands/growth_checker.js]: `;`

File: commands/growth_checker.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        },
+
+        get_dates = function (data, channel, cb) {
+            var range;
+            channel.dates = [];
+
+            if (!data.items[0]) {
+                app.log.error(channel.channel_id + ' cannot find youtube data.');
+                failed++;
+                return cb(null, {channel_id : channel.channel_id, error: 'cannot find youtube data'});
```
Comment (adin234) [inline commands/growth_checker.js]: don't we want to call `cb` here sending an error?

File: commands/growth_checker.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            }
+
+            range = moment.range(new Date(data.items[0].snippet.publishedAt), moment(new Date()).subtract(2, 'days'));
+            range.by('days', function (moment) {
+                channel.dates.push(moment.format('YYYY-MM-DD'));
+            });
```
Comment (adin234) [inline commands/growth_checker.js]: new line?

File: commands/growth_checker.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            return get_in_db(channel, cb);
+        },
+
+        get_in_db = function (channel, cb) {
+            return mysql.open(app.config.get('database').heartbeat)
+                .query('SELECT distinct(date) FROM channel_growth WHERE channel_id = ?', channel.channel_id, function(err, results) {
+                    if (err) {
+                        app.log.error(channel.channel_id + ' something went wrong.');
+                        failed++;
+                        return cb(null, {channel_id : channel.channel_id, error: 'SQL Error'});
```
Comment (adin234) [inline commands/growth_checker.js]: don't we want to call `cb` here sending an error?

File: commands/growth_checker.js [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+                    return get_difference(results, channel, cb);
+                });
+        },
+
+        get_difference = function (results, channel, cb) {
+            var diffs = 0;
+
+            results = results.map(function (e) {return moment(e.date).format('YYYY-MM-DD');});
+
+            diffs = channel.dates.filter(function(i) {
```
Comment (adin234) [inline commands/growth_checker.js]: quick tip: if your code will do many array operations, we do suggest using [lodash](https://lodash.com/)

File: commands/growth_checker.js [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        },
+
+        save_diffs = function (diffs, channel, cb) {
+            var values = [];
+            diffs.forEach(function (e) {
+                values.push([channel.channel_id, channel.network, e]);
+            });
+
+            return mysql.open(app.config.get('database').heartbeat)
+                .query('INSERT INTO jobs (channel_id, network, dates) VALUES ?', [values], function(err, results) {
```
Comment (adin234) [inline commands/growth_checker.js]: `function (err, result)`

File: commands/growth_checker.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+                    if (err) {
+                        app.log.error(channel.channel_id + ' something went wrong.');
+                        failed++;
+                        return cb(null, {channel_id : channel.channel_id, error: 'SQL Error'});
```
Comment (adin234) [inline commands/growth_checker.js]: don't we want to call `cb` here sending an error?

File: commands/growth_checker.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+                    }
+
+                    success++;
+                    return cb(null, channel.stats);
+                });
+        },
+
+        start = function () {
```
Comment (adin234) [inline commands/growth_checker.js]: we normally declare start at the **start** of the code

File: commands/growth_checker.js (hunk 13 above)
Comment (eshleebien) [inline commands/growth_checker.js]: rename jobs to channel_growth_jobs

File: commands/growth_checker.js (hunk 14 above)
Comment (eshleebien) [inline commands/growth_checker.js]: did you mean next? instead of cb?

File: commands/growth_checker.js (hunk 10 above)
Comment (rycastaneda) [inline commands/growth_checker.js]: there is a new line in line 104

File: commands/growth_checker.js (hunk 4 above)
Comment (rycastaneda) [inline commands/growth_checker.js]: line 62

File: commands/growth_checker.js [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+        },
+
+        process_rows = function (err, results) {
+            if (err) {
+                app.log.error('Something went wrong.');
+                return next();
+            }
+
+            async.map(results, get_youtube, function (err, result) {
+                app.log.info("result", result);
```
Comment (ravenjohn) [inline commands/growth_checker.js]: single quotes

File: commands/growth_checker.js [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+            diffs = _(channel.dates).difference(results).value();
+
+            channel.stats = {
+                channel : channel.channel_id,
+                db : results.length,
+                diffs : diffs.length,
+                dates : channel.dates.length
+            }
+
+            app.log.info("channel.stats", channel.stats);
```
Comment (ravenjohn) [inline commands/growth_checker.js]: single quotes

File: commands/growth_checker.js [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+var mysql  = require('anytv-node-mysql'),
+    cudl   = require('cuddle'),
+    moment = require('moment'),
+    async = require('async'),
+    _ = require('lodash'),
+    aws = require('aws-sdk'),
+
+    app = {};
+
+require('moment-range');
```
Comment (adin234) [inline commands/growth_checker.js]: you really meant to require this here without saving to any variable?

File: commands/growth_checker.js [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+
+        start = function () {
+            return get_channels();
+        },
+
+        get_channels = function () {
+            var query = 'SELECT channel_id, network \
+                         FROM mcn_channels\
+                         WHERE (linked_date != "" OR linked_date is not null)\
+                         AND (network != "")';
```
Comment (eshleebien) [inline commands/growth_checker.js]: and is_terminated != 1

**Comment by adin234** [comment]
hello, we have updated javascript standards just this weekend, we think it is best if we move the coding style of this repo to the updated [standards](https://github.com/anyTV/JS-conventions)

**Pull request merged**